src/ngx_http_lua_worker_thread.c

```diff
@@ -0,0 +1,356 @@
+{
+    task_ctx_t* ctx = NULL;
+    if (ctxpool->next == NULL) {
+        ctx = (task_ctx_t*)calloc(sizeof(task_ctx_t), 1);
+        lua_State* vm = luaL_newstate();
+        ctx->vm = vm;
+        luaL_openlibs(ctx->vm);
+
+        lua_getglobal(L, "package");
+        lua_getglobal(vm, "package");
+
+        // copy package.path
+        lua_getfield(L, -1, "path");
+        size_t path_len;
+        const char* path = lua_tolstring(L, -1, &path_len);
+        lua_pushlstring(vm, path, path_len);
+        lua_setfield(vm, -2, "path");
+        lua_pop(L, 1);
+
+        // copy package.cpath
+        lua_getfield(L, -1, "cpath");
+        size_t cpath_len;
+        const char* cpath = lua_tolstring(L, -1, &cpath_len);
+        lua_pushlstring(vm, cpath, cpath_len);
+        lua_setfield(vm, -2, "cpath");
+        lua_pop(L, 1);
+
+        // remove the "package" table
+        lua_pop(L, 1);
+        lua_pop(vm, 1);
+
+        ngx_thread_mutex_create(&ctx->mutex, ngx_cycle->log);
+    } else {
+        ctx = ctxpool->next;
+        ctxpool->next = ctxpool->next->next;
+    }
+    return ctx;
+}
+
+static void
+put_task_ctx(task_ctx_t* ctx)
+{
+    ctx->next = ctxpool->next;
+    ctxpool->next = ctx;
+    /*  clear Lua stack */
+    lua_settop(ctx->vm, 0);
+}
+
+static int
+lauxh_xcopy(lua_State *from, lua_State *to, int idx, const int allow_nil)
+{
+    switch(lua_type(from, idx))
+    {
+        case LUA_TBOOLEAN:
+            lua_pushboolean(to, lua_toboolean(from, idx));
+            return LUA_TBOOLEAN;
+
+        case LUA_TLIGHTUSERDATA:
+            lua_pushlightuserdata(to, lua_touserdata(from, idx));
+            return LUA_TLIGHTUSERDATA;
+
+        case LUA_TNUMBER:
+            lua_pushnumber(to, lua_tonumber(from, idx));
+            return LUA_TNUMBER;
+
+        case LUA_TSTRING: {
+            size_t len = 0;
+            const char *str = lua_tolstring(from, idx, &len);
+            lua_pushlstring(to, str, len);
+            return LUA_TSTRING;
+        }
+
+        case LUA_TTABLE:
+            lua_newtable(to);
+            // to positive number
+            if(idx < 0) {
+                idx = lua_gettop(from) + idx + 1;
+            }
+            lua_pushnil(from);
+            while(lua_next(from, idx))
+            {
+                if(lauxh_xcopy(from, to, -2, 0) != LUA_TNONE)
+                {
+                    if(lauxh_xcopy(from, to, -1, 0) != LUA_TNONE) {
+                        lua_rawset(to, -3);
+                    }
+                    else {
+                        lua_pop(to, 1);
+                    }
+                }
+                lua_pop(from, 1);
+            }
+            return LUA_TTABLE;
+
+        case LUA_TNIL:
+            if(allow_nil) {
+                lua_pushnil(to);
+                return LUA_TNIL;
+            }
+            // fall through
+
+        // ignore unsupported values
+        // LUA_TNONE
+        // LUA_TFUNCTION
+        // LUA_TUSERDATA
+        // LUA_TTHREAD

```

Comment:
Please don't use C++ style line comments. That's against the nginx coding style.